### PR TITLE
Ticket2559: Create configurations for forwarder from blockserver

### PR DESCRIPTION
**Pull request by DominicOram**
### Description of work

See https://github.com/ISISComputingGroup/IBEX/issues/2559

### To test

* Start epics to kafka forwarder
* Start the BlockServerToKafka, pointing it at the correct data and config topics and supplying an instrument prefix with a running blockserver
* Confirm PV data is being put into the data topic based on the blocks
* Change config
* Confirm data for new blocks is in data topic

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Has the author taken into account the multi-threaded nature of the code?
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [ ] Has the [manual system tests spreadsheet](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Manual-system-tests) been updated?

### Functional Tests

- [ ] Do changes function as described? Add comments below that describe the tests performed.

### Final steps

- [ ] Reviewer has updated the submodule in the main EPICS repo? See **Reviewing work for the subModules of EPICS** in the [Git workflow](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Git-workflow) page for details.
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: BlockServerToKafka/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+__author__ = 'ffv81422'
```
Comment (DominicOram) [inline BlockServerToKafka/__init__.py]: Remove this

File: BlockServerToKafka/forwarder_config.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    def __init__(self, topic, using_v4=False, schema="f142"):
+        self.schema = schema
+        self.topic = topic
+        self.using_v4 = using_v4
+
+    def _get_converter(self):
+        return {"schema": self.schema, "topic": self.topic}
+
+    def _create_stream(self, blk):
+        stream = dict()
```
Comment (DominicOram) [inline BlockServerToKafka/forwarder_config.py]: Change to {}

File: BlockServerToKafka/forwarder_config.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        stream["channel"] = blk
+        stream["converter"] = self._get_converter()
+        if not self.using_v4:
+            stream["channel_provider_type"] = "ca"
+        return stream
+
+    def create_forwarder_configuration(self, pvs):
+        output_dict = dict()
```
Comment (DominicOram) [inline BlockServerToKafka/forwarder_config.py]: Change to {}

File: BlockServerToKafka/forwarder_config.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+        output_dict["cmd"] = "add"
+
+        streams = []
```
Comment (DominicOram) [inline BlockServerToKafka/forwarder_config.py]: List comprehension

File: BlockServerToKafka/forwarder_config.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        return json.dumps(output_dict)
+
+    def remove_forwarder_configuration(self, pvs):
+        output_list = []
+        for pv in pvs:
+            out_dict = dict()
+            out_dict["cmd"] = "stop_channel"
+            out_dict["channel"] = pv
+            output_list.append(json.dumps(out_dict))
+        return output_list
```
Comment (DominicOram) [inline BlockServerToKafka/forwarder_config.py]: new line

**Comment by DominicOram** [comment]
ADD COMMENTS THROUGHOUT!!!!!!

File: BlockServerToKafka/kafka_producer.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+# You should have received a copy of the Eclipse Public License v1.0
+# along with this program; if not, you can obtain a copy from
+# https://www.eclipse.org/org/documents/epl-v10.php or
+# http://opensource.org/licenses/eclipse-1.0.php
+from forwarder_config import ForwarderConfig
+
+
+class Producer():
+    def __init__(self, server, config_topic, data_topic):
+        from kafka import KafkaProducer
```
Comment (DominicOram) [inline BlockServerToKafka/kafka_producer.py]: Import outside constructor

File: BlockServerToKafka/test_modules/__init__.py [hunk 7]
```diff
@@ -0,0 +1 @@
+__author__ = 'ffv81422'
```
Comment (DominicOram) [inline BlockServerToKafka/test_modules/__init__.py]: Remove this

File: BlockServerToKafka/test_modules/kafka_tests.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    def setUp(self):
+        self.kafka_forwarder = ForwarderConfig(self.test_topic, False, self.test_schema)
+        self.config_with_one_block = [self.test_block_1]
+        self.config_with_two_blocks = [self.test_block_1, self.test_block_2]
+
+    def test_WHEN_new_forwarder_config_created_THEN_returns_valid_JSON(self):
+        output = self.kafka_forwarder.create_forwarder_configuration(self.config_with_one_block)
+        try:
+            json.loads(output)
+        except Exception as e:
```
Comment (DominicOram) [inline BlockServerToKafka/test_modules/kafka_tests.py]: Specific exception, see if we can do this nicer in unittest

File: BlockServerToKafka/test_modules/kafka_tests.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+    def test_GIVEN_schema_and_topic_WHEN_forwarder_config_created_THEN_returns_JSON_containing_schema_and_topic(self):
+        raw_output = self.kafka_forwarder.create_forwarder_configuration(self.config_with_one_block)
+        output = json.loads(raw_output)
+        self.assertNotEqual(len(output["streams"]), 0)
+        for stream in output["streams"]:
+            self.assertNotEqual(len(stream["converter"]), 0)
+            self.assertTrue("schema" in stream["converter"])
+            self.assertTrue("topic" in stream["converter"])
+            self.assertEqual(self.test_schema, stream["converter"]["schema"])
```
Comment (DominicOram) [inline BlockServerToKafka/test_modules/kafka_tests.py]: Expected at the front

File: BlockServerToKafka/test_modules/kafka_tests.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        self.assertNotEqual(len(output["streams"]), 0)
+        for stream in output["streams"]:
+            self.assertEqual(type(stream), dict)
+            self.assertTrue("channel" in stream)
+            self.assertTrue("converter" in stream)
+
+    def test_GIVEN_schema_and_topic_WHEN_forwarder_config_created_THEN_returns_JSON_containing_schema_and_topic(self):
+        raw_output = self.kafka_forwarder.create_forwarder_configuration(self.config_with_one_block)
+        output = json.loads(raw_output)
+        self.assertNotEqual(len(output["streams"]), 0)
```
Comment (DominicOram) [inline BlockServerToKafka/test_modules/kafka_tests.py]: Expected at front

File: BlockServerToKafka/forwarder_config.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+    def _get_converter(self):
+        return {"schema": self.schema, "topic": self.topic}
+
+    def _create_stream(self, blk):
+        stream = dict()
+        stream["channel"] = blk
+        stream["converter"] = self._get_converter()
+        if not self.using_v4:
+            stream["channel_provider_type"] = "ca"
```
Comment (DominicOram) [inline BlockServerToKafka/forwarder_config.py]: Explicit with pva if v4 (check that forwarder is happy with this)

File: BlockServerToKafka/test_modules/kafka_tests.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        output = json.loads(raw_output)
+        stream = output["streams"][0]
+        self.assertEqual(stream["channel"], self.test_block_1)
+
+    def test_GIVEN_configuration_with_two_blocks_WHEN_new_forwarder_config_created_THEN_returns_both_block_pv_string(self):
+        raw_output = self.kafka_forwarder.create_forwarder_configuration(self.config_with_two_blocks)
+        output = json.loads(raw_output)
+
+        blks = [self.test_block_1, self.test_block_2]
+        for i in range(2):
```
Comment (DominicOram) [inline BlockServerToKafka/test_modules/kafka_tests.py]: Use enumerate

Comment (DominicOram) [reply BlockServerToKafka/test_modules/kafka_tests.py]: Think of how to do this without assuming json ordered

File: BlockServerToKafka/test_modules/kafka_tests.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+            stream = output["streams"][i]
+            self.assertEqual(stream["channel"], blks[i])
+
+    def test_WHEN_removed_old_forwarder_THEN_JSON_returns_valid(self):
+        output = self.kafka_forwarder.remove_forwarder_configuration(self.config_with_one_block)
+        for js in output:
+            try:
```
Comment (DominicOram) [inline BlockServerToKafka/test_modules/kafka_tests.py]: Same as above exception bit

File: BlockServerToKafka/test_modules/kafka_tests.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+            output = json.loads(js)
+            self.assertTrue("channel" in output)
+            self.assertEqual(output["channel"], self.test_block_1)
+
+    def test_GIVEN_configuration_with_two_blocks_WHEN_removed_old_forwarder_THEN_returns_JSON_containing_both_block_pv_string(self):
+        raw_output = self.kafka_forwarder.remove_forwarder_configuration(self.config_with_two_blocks)
+        self.assertTrue(len(raw_output) == 2)
+
+        blks = [self.test_block_1, self.test_block_2]
+        for i in range(2):
```
Comment (DominicOram) [inline BlockServerToKafka/test_modules/kafka_tests.py]: Same as above for enumerate/json ordering

File: BlockServerToKafka/test_modules/kafka_tests.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+            output = json.loads(raw_output[i])
+            self.assertTrue("channel" in output)
+            self.assertEqual(output["channel"], blks[i])
```
Comment (DominicOram) [inline BlockServerToKafka/test_modules/kafka_tests.py]: new line

File: BlockServerToKafka/block_server_monitor.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+    def __init__(self, address, pvprefix, producer):
+        self.PVPREFIX = pvprefix
+        self.address = address
+        self.channel = CaChannel()
+        self.producer = producer
+        self.last_pv = []
+        try:
+            self.channel.searchw(self.address)
+        except CaChannelException:
+            print ("Unable to find pv " + self.address)
```
Comment (DominicOram) [inline BlockServerToKafka/block_server_monitor.py]: Use format, no space after print

Comment (DominicOram) [reply BlockServerToKafka/block_server_monitor.py]: Print and log

File: BlockServerToKafka/block_server_monitor.py [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        data = str(bytearray(epics_args['pv_value'])).replace("\x00", "")
+        data = utilities.dehex_and_decompress(data)
+        blocks = json.loads(data)
+
+        pvs = [self.block_name_to_pv_name(blk) for blk in blocks]
+        if pvs != self.last_pv:
+            print("Configuration changed to: {}".format(pvs))
+            self.producer.remove_config(self.last_pv)
+            self.producer.add_config(pvs)
+            self.last_pv = pvs
```
Comment (DominicOram) [inline BlockServerToKafka/block_server_monitor.py]: New line

File: BlockServerToKafka/block_server_monitor.py [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+from BlockServer.core.macros import BLOCK_PREFIX
+
+
+class BlockServerMonitor:
+    def __init__(self, address, pvprefix, producer):
+        self.PVPREFIX = pvprefix
+        self.address = address
+        self.channel = CaChannel()
+        self.producer = producer
+        self.last_pv = []
```
Comment (DominicOram) [inline BlockServerToKafka/block_server_monitor.py]: pvS

File: BlockServerToKafka/block_server_monitor.py [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+            ca.DBE_VALUE,
+            self.update,
+            None)
+        self.channel.pend_event()
+
+    def block_name_to_pv_name(self, blk):
+        return '{}{}{}'.format(self.PVPREFIX, BLOCK_PREFIX, blk.upper())
+
+    def update(self, epics_args, user_args):
+        data = str(bytearray(epics_args['pv_value'])).replace("\x00", "")
```
Comment (DominicOram) [inline BlockServerToKafka/block_server_monitor.py]: Comment and look into why this is. Can we get NORD?

File: BlockServerToKafka/block_server_monitor.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+        data = utilities.dehex_and_decompress(data)
+        blocks = json.loads(data)
+
+        pvs = [self.block_name_to_pv_name(blk) for blk in blocks]
+        if pvs != self.last_pv:
+            print("Configuration changed to: {}".format(pvs))
```
Comment (DominicOram) [inline BlockServerToKafka/block_server_monitor.py]: Change to print and log

**Comment by mattclarke** [comment]
Looks good, just one thing though: can you make sure the docstrings match what is described in https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Python-conventions

I'll highlight some, but not all, of the differences

File: BlockServerToKafka/block_server_monitor.py [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+from CaChannel import CaChannelException
+from server_common.utilities import dehex_and_decompress, print_and_log
+import ca
+import json
+from BlockServer.core.macros import BLOCK_PREFIX
+from threading import RLock
+
+
+class BlockServerMonitor:
+    """ Class that monitors the blockserver to see when the config has changed.
```
Comment (mattclarke) [inline BlockServerToKafka/block_server_monitor.py]: Newline after  opening """

File: BlockServerToKafka/block_server_monitor.py [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+            0,
+            ca.DBE_VALUE,
+            self.update,
+            None)
+        self.channel.pend_event()
+
+    def block_name_to_pv_name(self, blk):
+        """ Converts a block name to a PV name by adding the prefixes.
+        Args:
+            blk (string): the name of the block
```
Comment (mattclarke) [inline BlockServerToKafka/block_server_monitor.py]: We are now going for full sentences style:

blk (string): The name of the block.

File: BlockServerToKafka/block_server_monitor.py [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+        Returns:
+            (string) the associated PV name
+        """
```
Comment (mattclarke) [inline BlockServerToKafka/block_server_monitor.py]: Should be:

string : The associated PV name.

File: BlockServerToKafka/block_server_monitor.py [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+        return '{}{}{}'.format(self.PVPREFIX, BLOCK_PREFIX, blk)
+
+    def convert_to_string(self, pv_array):
+        """ Converts from byte array to string and removes null characters
+        Args:
+            pv_array(bytearray): The byte array of PVs.
```
Comment (mattclarke) [inline BlockServerToKafka/block_server_monitor.py]: insert space before bracket

File: BlockServerToKafka/block_server_monitor.py [hunk 25]
```diff
@@ -0,0 +1,92 @@
...
+        """ Converts a block name to a PV name by adding the prefixes.
+        Args:
+            blk (string): the name of the block
+        Returns:
+            (string) the associated PV name
+        """
+        return '{}{}{}'.format(self.PVPREFIX, BLOCK_PREFIX, blk)
+
+    def convert_to_string(self, pv_array):
+        """ Converts from byte array to string and removes null characters
```
Comment (mattclarke) [inline BlockServerToKafka/block_server_monitor.py]: Needs a full stop.

**Comment by rerpha** [comment]
may want to add a try/catch statement where the kafkaproducer is set up, if it can't find a broker it will crash and throw an exception which is defined in the kafka library.

**Comment by rerpha** [comment]
note: the forwarder will only throw an error if `auto.create.topics.enable` is set to false in the broker configuration. If not it will create a new topic with the name specified. 

[http://kafka.apache.org/documentation.html#configuration](url)

**Pull request merged**